xbmc/guilib/GUITextLayout.cpp

```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
   for (unsigned int i=0; i<lines.size(); i++)
   {
     CGUIString &line = lines[i];
+	unsigned int lineLength = line.m_text.size();
+	std::wstring logicalText;
+    vecText style;
 
-    // reserve enough space in the flipped text
-    vecText flippedText;
-    flippedText.reserve(line.m_text.size());
+    logicalText.reserve(lineLength);
+    style.reserve(lineLength);
 
-    character_t sectionStyle = 0xffff0000; // impossible to achieve
-    std::wstring sectionText;
+    // Separate the text and style for the input styled text
     for (const auto& it : line.m_text)
     {
-      character_t style = it & 0xffff0000;
-      if (style != sectionStyle)
-      {
-        if (!sectionText.empty())
-        { // style has changed, bidi flip text
-          std::wstring sectionFlipped = BidiFlip(sectionText, forceLTRReadingOrder);
-          for (unsigned int j = 0; j < sectionFlipped.size(); j++)
-            flippedText.push_back(sectionStyle | sectionFlipped[j]);
-        }
-        sectionStyle = style;
-        sectionText.clear();
-      }
-      sectionText.push_back((wchar_t)(it & 0xffff));
+      logicalText.push_back((wchar_t)(it & 0xffff));
+      style.push_back(it & 0xffff0000);
     }
+	
+	// Allocate memory for logical to visual map and call bidi
+    int* logicalToVisualMap = (int*)malloc((lineLength + 1) * sizeof(int));
```

Comment:
Why malloc? is there a reason not to use `new`?

`int* logicalToVisualMap = new int[lineLength + 1];`